File: LeetCode Problems/101 to 200/105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    unordered_map<int, int> mapValIdx;
    int preIdx = 0;
    TreeNode* dfs_fun(int left, int right, vector<int> &preorder) {
        if (left > right) return nullptr;
        TreeNode *root = new TreeNode(preorder[preIdx++]);
        int mid = mapValIdx[root->val];
        root->left = dfs_fun(left, mid - 1, preorder);
        root->right = dfs_fun(mid + 1, right, preorder);
        return root;
    }
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
         int n = inorder.size();
        for (int i = 0; i < n; ++i) mapValIdx[inorder[i]] = i;
        return dfs_fun(0, n - 1, preorder);
    }
};

File: LeetCode Problems/101 to 200/105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    HashMap<Integer, Integer> mapVal2Idx = new HashMap<>();
    int preIdx = 0;
    
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        int n = inorder.length;
        for (int i = 0; i < n; ++i) mapVal2Idx.put(inorder[i], i);
        return dfs(0, n - 1, preorder);
    }
    
    TreeNode dfs(int left, int right, int[] preorder) {
        if (left > right) return null;
        TreeNode root = new TreeNode(preorder[preIdx++]);
        int mid = mapVal2Idx.get(root.val);
        root.left = dfs(left, mid - 1, preorder);
        root.right = dfs(mid + 1, right, preorder);
        return root;
    }
}